src/Extensions/Hosting.Services.Web/Middlewares/ObsoleteCorrelationHeadersMiddleware.cs

```diff
@@ -0,0 +1,132 @@
+using Microsoft.Omex.Extensions.Abstractions.Activities;
+
+namespace Microsoft.Omex.Extensions.Hosting.Services.Web.Middlewares
+{
+	/// <summary>
+	/// Adds support of old correlation tracking logic that might be used by other services
+	/// </summary>
+	/// <remarks>
+	/// This middleware should be removed after all services moved to use <see cref="Activity"/>
+	/// </remarks>
+	[Obsolete("Use it only if you need to comunicate with services that use old correlation", false)]
+	internal class ObsoleteCorrelationHeadersMiddleware
+	{
+		/// <summary>
+		/// Invoke middleware
+		/// </summary>
+		public Task InvokeAsync(HttpContext context)
+		{
+			ExtractCorrelationFromRequest(context.Request);
+			context.Response.OnStarting(SetCorrelationHeadersToResponce, context.Response);
+			return Task.CompletedTask;
+		}
+
+		private Task SetCorrelationHeadersToResponce(object state)
+		{
+			HttpResponse response = (HttpResponse)state;
+			Guid? oldCorrelation = Activity.Current?.GetObsoleteCorrelationId();
+			if (oldCorrelation.HasValue)
+			{
+				response.Headers.Add(CorrelationHeader, oldCorrelation.ToString());
+			}
+
+			return Task.CompletedTask;
+		}
+
+		private void ExtractCorrelationFromRequest(HttpRequest request)
+		{
+			Activity? activity = Activity.Current;
+			if (activity == null)
+			{
+				return;
+			}
+
+			Guid? oldCorrelation = ExtractCorrelationId(request);
+			if (oldCorrelation.HasValue)
+			{
+				activity.SetObsoleteCorrelationId(oldCorrelation.Value);
+			}
+
+			IQueryCollection dataSources = request.Query;
+
+			uint? transactionId = ParseUint(ExtractParameter(dataSources, s_transactionsNames));
+			if (transactionId.HasValue)
+			{
+				activity.SetObsoleteTransactionId(transactionId.Value);
+			}
+			else
+			{
+				return;
+			}
+		}
+
+		private static Guid? ExtractCorrelationId(HttpRequest request) =>
+			!IsClientRequest(request)
+			&& Guid.TryParse(ExtractParameter(request.Query, s_correlationIdNames), out Guid correlation)
+				? correlation
+				: (Guid?)null;
+
+		private static string? ExtractParameter(IQueryCollection dataSources, IEnumerable<string> names)
+		{
+			foreach (string name in names)
+			{
+				string? value = ExtractParameter(dataSources, name);
+				if (!string.IsNullOrWhiteSpace(value))
+				{
+					return value;
+				}
+			}
+
+			return null;
+		}
+
+		private static string? ExtractParameter(IQueryCollection dataSources, string name) =>
+			dataSources.TryGetValue(name, out StringValues value) && value.Count > 0
+				? value.FirstOrDefault(s => !string.IsNullOrWhiteSpace(s))
+				: null;
+
+		/// <summary>
+		/// Checks if the context is for a request that contains identifiers indicating that the request originated from an Office client or not
+		/// </summary>
+		private static bool IsClientRequest(HttpRequest request) =>
+			request.Headers.ContainsKey(OfficeClientVersionHeader)
+				? true
+				: request.Query.Count < 1 // Don't check empty params
```

Comment:
>request.Query.Count < 1  [](start = 6, length = 24)

Would it be more readable to say

request.Query.Count == 0?

Reply:
Sure can do it like this also, I think I was worried about the case if Count will return negative value :)